Ifpack2:  Build error mixing complex<float> and complex<double> in Thyra_Ifpack2PreconditionerFactory_def.hpp

## Bug Report
@trilinos/Ifpack2 @trilinos/Thyra

### Description
What went wrong?  What should have happened?  Do you have an idea what might
fix things?  How has this bug affected you?  What are you trying to accomplish?
What specifically must we demonstrate to call this work complete?

### Steps to Reproduce
1. SHA1:  db2b7fb220f3853fd7bd325775a3c51f908b54f3 (current develop)

Build error is:
```
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/tpetra/core/src/Tpetra_MixedScalarMultiplyOp.hpp:125:35: error: no viable conversion from 'complex<double>' to 'complex<float>'
      op_->apply (*X_, *Y_, mode, alpha, beta);
                                  ^~~~~
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/tpetra/core/src/Tpetra_MixedScalarMultiplyOp.hpp:94:5: note: in instantiation of member function 'Tpetra::MixedScalarMultiplyOp<std::complex<double>, std::complex<float>, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::apply' requested here
    MixedScalarMultiplyOp (const Teuchos::RCP<const op_type>& A) :
    ^
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/ifpack2/adapters/thyra/Thyra_Ifpack2PreconditionerFactory_def.hpp:252:42: note: in instantiation of member function 'Tpetra::MixedScalarMultiplyOp<std::complex<double>, std::complex<float>, int, long, Kokkos::Compat::KokkosDeviceWrapperNode<Kokkos::Serial, Kokkos::HostSpace> >::MixedScalarMultiplyOp' requested here
    RCP<TpetraLinOp> wrappedOp = rcp(new Tpetra::MixedScalarMultiplyOp<scalar_type,half_scalar_type,local_ordinal_type,global_ordinal_type,node_type>(concretePrecOpHalf));
                                         ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1051:12: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'std::complex<double>' to 'const std::complex<float> &' for 1st argument
    struct complex<float>
           ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1051:12: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'std::complex<double>' to 'std::complex<float> &&' for 1st argument
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1056:26: note: candidate constructor not viable: no known conversion from 'std::complex<double>' to 'std::complex<float>::_ComplexT' (aka '_Complex float') for 1st argument
      _GLIBCXX_CONSTEXPR complex(_ComplexT __z) : _M_value(__z) { }
                         ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1058:26: note: candidate constructor not viable: no known conversion from 'std::complex<double>' to 'float' for 1st argument
      _GLIBCXX_CONSTEXPR complex(float __r = 0.0f, float __i = 0.0f)
                         ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1068:35: note: explicit constructor is not a candidate
      explicit _GLIBCXX_CONSTEXPR complex(const complex<double>&);
                                  ^
/projects/sierra/linux_rh7/SDK/compilers/gcc/7.2.0-RHEL7/lib/gcc/x86_64-pc-linux-gnu/7.2.0/../../../../include/c++/7.2.0/complex:1069:35: note: explicit constructor is not a candidate
      explicit _GLIBCXX_CONSTEXPR complex(const complex<long double>&);
                                  ^
/scratch/vebruni/src/code/objs/tpls/spack/spack/.spack/stage/spack-stage-trilinos-2021.08.16-qi2tfvfw2j3jdwl6pqwi5jfail4q2m6y/spack-src/packages/tpetra/core/src/Tpetra_Operator.hpp:123:19: note: passing argument to parameter 'alpha' here
           Scalar alpha = Teuchos::ScalarTraits<Scalar>::one(),
```

--- a/packages/teuchos/core/src/Teuchos_as.hpp
+++ b/packages/teuchos/core/src/Teuchos_as.hpp
@@ -2576,6 +2576,21 @@ public:
   }
 };
 
+
+//! Convert std::complex<double> to std::complex<float>.
+template<>
+class ValueTypeConversionTraits<std::complex<float>, std::complex<double> > {
+public:
+  inline static std::complex<float> convert (const std::complex<double> t) {
+    // Let double->float conversion handle real and imaginary part.
+    return std::complex<float> (as<float> (t.real ()), as<float> (t.imag ()));
+  }
+  static std::complex<float> safeConvert (const std::complex<double> t) {
+    // Let double->float conversion handle real and imaginary part.
+    return std::complex<float> (asSafe<float> (t.real ()), asSafe<float> (t.imag ()));
+  }
+};
+
 #endif // HAVE_TEUCHOS_COMPLEX
 
 //

--- a/packages/teuchos/core/test/TypeConversions/TypeConversions_UnitTest.cpp
+++ b/packages/teuchos/core/test/TypeConversions/TypeConversions_UnitTest.cpp
@@ -1057,6 +1057,115 @@ UNIT_TEST_GROUP_SIGNED_INTEGER( long_long_type )
 //UNIT_TEST_GROUP_ANY_INTEGER( unsigned long long )
 UNIT_TEST_GROUP_ANY_INTEGER( unsigned_long_long_type )
 
+
+#ifdef HAVE_TEUCHOS_COMPLEX
+
+#include <complex>
+
+//
+// Tests for conversions involving complex types.
+//
+TEUCHOS_UNIT_TEST( asSafe, complexToComplex ) {
+  using Teuchos::as;
+  using Teuchos::asSafe;
+
+  const std::complex<float>  iF = std::complex<float>(0,1);
+  const std::complex<double> iD = std::complex<double>(0,1);
+
+  const std::complex<float> minF = -std::numeric_limits<float>::max ();
+  const std::complex<float> minImagF = -iF*std::numeric_limits<float>::max ();
+  const std::complex<float> minusOneF = -1;
+  const std::complex<float> minusOneImagF = -iF;
+  const std::complex<float> maxF = std::numeric_limits<float>::max ();
+  const std::complex<float> maxImagF = iF*std::numeric_limits<float>::max ();
+
+  const std::complex<double> minD = -std::numeric_limits<double>::max ();
+  const std::complex<double> minImagD = -iD*std::numeric_limits<double>::max ();
+  const std::complex<double> minusOneD = -1;
+  const std::complex<double> minusOneImagD = -iD;
+  const std::complex<double> maxD = std::numeric_limits<double>::max ();
+  const std::complex<double> maxImagD = iD*std::numeric_limits<double>::max ();
+
+  std::complex<float> valF = 0;
+
+  std::complex<double> valD = 0;
+
+  //
+  // Test std::complex<float> -> std::complex<float> conversions.
+  //
+
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (minF));
+  TEST_EQUALITY_CONST(valF, minF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (minF));
+  TEST_EQUALITY_CONST(valF, minF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (minImagF));
+  TEST_EQUALITY_CONST(valF, minImagF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (minImagF));
+  TEST_EQUALITY_CONST(valF, minImagF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (maxF));
+  TEST_EQUALITY_CONST(valF, maxF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (maxF));
+  TEST_EQUALITY_CONST(valF, maxF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (maxImagF));
+  TEST_EQUALITY_CONST(valF, maxImagF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (maxImagF));
+  TEST_EQUALITY_CONST(valF, maxImagF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (minusOneF));
+  TEST_EQUALITY_CONST(valF, minusOneF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (minusOneF));
+  TEST_EQUALITY_CONST(valF, minusOneF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (minusOneImagF));
+  TEST_EQUALITY_CONST(valF, minusOneImagF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (minusOneImagF));
+  TEST_EQUALITY_CONST(valF, minusOneImagF);
+
+
+  //
+  // Test std::complex<double> -> std::complex<double> conversions.
+  //
+
+  TEST_NOTHROW(valD = asSafe<std::complex<double> > (minD));
+  TEST_EQUALITY_CONST(valD, minD);
+  TEST_NOTHROW(valD = as<std::complex<double> > (minD));
+  TEST_EQUALITY_CONST(valD, minD);
+  TEST_NOTHROW(valD = asSafe<std::complex<double> > (minImagD));
+  TEST_EQUALITY_CONST(valD, minImagD);
+  TEST_NOTHROW(valD = as<std::complex<double> > (minImagD));
+  TEST_EQUALITY_CONST(valD, minImagD);
+  TEST_NOTHROW(valD = asSafe<std::complex<double> > (maxD));
+  TEST_EQUALITY_CONST(valD, maxD);
+  TEST_NOTHROW(valD = as<std::complex<double> > (maxD));
+  TEST_EQUALITY_CONST(valD, maxD);
+  TEST_NOTHROW(valD = asSafe<std::complex<double> > (maxImagD));
+  TEST_EQUALITY_CONST(valD, maxImagD);
+  TEST_NOTHROW(valD = as<std::complex<double> > (maxImagD));
+  TEST_EQUALITY_CONST(valD, maxImagD);
+  TEST_NOTHROW(valD = asSafe<std::complex<double> > (minusOneD));
+  TEST_EQUALITY_CONST(valD, minusOneD);
+  TEST_NOTHROW(valD = as<std::complex<double> > (minusOneD));
+  TEST_EQUALITY_CONST(valD, minusOneD);
+  TEST_NOTHROW(valD = asSafe<std::complex<double> > (minusOneImagD));
+  TEST_EQUALITY_CONST(valD, minusOneImagD);
+  TEST_NOTHROW(valD = as<std::complex<double> > (minusOneImagD));
+  TEST_EQUALITY_CONST(valD, minusOneImagD);
+
+  //
+  // Test std::complex<double> -> std::complex<float> conversions.
+  //
+
+  TEST_NOTHROW(valF = as<std::complex<float> > (minusOneD));
+  TEST_EQUALITY_CONST(valF, minusOneF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (minusOneD));
+  TEST_EQUALITY_CONST(valF, minusOneF);
+  TEST_NOTHROW(valF = as<std::complex<float> > (minusOneImagD));
+  TEST_EQUALITY_CONST(valF, minusOneImagF);
+  TEST_NOTHROW(valF = asSafe<std::complex<float> > (minusOneImagD));
+  TEST_EQUALITY_CONST(valF, minusOneImagF);
+
+}
+
+#endif // HAVE_TEUCHOS_COMPLEX
+
 } // namespace (anonymous)
 
 

--- a/packages/tpetra/core/src/Tpetra_MixedScalarMultiplyOp.hpp
+++ b/packages/tpetra/core/src/Tpetra_MixedScalarMultiplyOp.hpp
@@ -122,7 +122,7 @@ namespace Tpetra {
       }
 
       Tpetra::deep_copy(*X_, X);
-      op_->apply (*X_, *Y_, mode, alpha, beta);
+      op_->apply (*X_, *Y_, mode, Teuchos::as<OpScalar>(alpha), Teuchos::as<OpScalar>(beta));
       Tpetra::deep_copy(Y, *Y_);
     }
 

